python_modules/dagster/dagster/_core/definitions/external_asset_graph.py

```diff
@@ -383,3 +384,24 @@ def split_asset_keys_by_repository(
                 asset_key
             )
         return list(asset_keys_by_repo.values())
+
+
+def _warn_on_duplicate_nodes(
+    node_pairs: Sequence[Tuple[RepositoryHandle, "ExternalAssetNode"]],
+    execution_type: AssetExecutionType,
+) -> None:
+    repo_handles_by_asset_key: DefaultDict[AssetKey, List[RepositoryHandle]] = defaultdict(list)
+    for repo_handle, node in node_pairs:
+        repo_handles_by_asset_key[node.asset_key].append(repo_handle)
+
+    duplicates = {k: v for k, v in repo_handles_by_asset_key.items() if len(v) > 1}
+    duplicate_lines = []
+    for asset_key, repo_handles in duplicates.items():
+        locations = [repo_handle.code_location_origin.location_name for repo_handle in repo_handles]
+        duplicate_lines.append(f"  {asset_key.to_string()}: {locations}")
+    duplicate_str = "\n".join(duplicate_lines)
+    if duplicates:
+        warnings.warn(
+            f"Found {execution_type.value} nodes for some asset keys in multiple code locations."
+            f" Only one {execution_type.value} node is allowed per asset key. Duplicates:\n {duplicate_str}"
+        )
```

Comment:
echoing a concern raised in discussions else where so its recorded here 

I think this will only show up in the daemon / webserver, which a user might catch in local dev but in a deployed contex is unlikely to be observed. Improving that effectively requires some UI surface area to highlight these global asset graph problems, much like we have talked about in the past for cycles.

Reply:
Raising this here: https://dagsterlabs.slack.com/archives/C03CCE471E0/p1708978938397619